overlord/snapstate/autorefresh_gating.go

```diff
@@ -33,6 +36,269 @@ import (
+	}
+	return fst.ModTime(), nil
+}
+
+type holdState struct {
+	// FirstHeld keeps the time when the given snap was first held for refresh by a gating snap.
+	FirstHeld time.Time `json:"first-held"`
+	// HoldUntil stores the desired end time for holding.
+	HoldUntil time.Time `json:"hold-until"`
+}
+
+func refreshGating(st *state.State) (map[string]map[string]*holdState, error) {
+	// held snaps -> holding snap(s) -> first-held/hold-until time
+	var gating map[string]map[string]*holdState
+	err := st.Get("snaps-hold", &gating)
+	if err != nil && err != state.ErrNoState {
+		return nil, fmt.Errorf("internal error: cannot get snaps-hold: %v", err)
+	}
+	if err == state.ErrNoState {
+		return make(map[string]map[string]*holdState), nil
+	}
+	return gating, nil
+}
+
+// HoldDurationError contains the that error prevents requested hold, along with
+// hold time that's left (if any).
+type HoldDurationError struct {
+	Err          error
+	DurationLeft time.Duration
+}
+
+func (h *HoldDurationError) Error() string {
+	return h.Err.Error()
+}
+
+// HoldError contains the details of snaps that cannot to be held.
+type HoldError struct {
+	SnapsInError map[string]HoldDurationError
+}
+
+func (h *HoldError) Error() string {
+	l := []string{""}
+	for _, e := range h.SnapsInError {
+		l = append(l, e.Error())
+	}
+	return fmt.Sprintf("cannot hold some snaps:%s", strings.Join(l, "\n - "))
+}
+
+func maxAllowedPostponement(gatingSnap, affectedSnap string) time.Duration {
+	if affectedSnap == gatingSnap {
+		return maxPostponement - maxPostponementBuffer
+	}
+	return maxOtherHoldDuration
+}
+
+// holdDurationLeft computes the maximum duration that's left for holding a refresh
+// given current time, last refresh time, time when snap was first held, maximum
+// duration allowed for the given snap and maximum overall postponement allowed by
+// snapd.
+func holdDurationLeft(now time.Time, lastRefresh, firstHeld time.Time, maxDuration, maxPostponement time.Duration) time.Duration {
+	d1 := firstHeld.Add(maxDuration).Sub(now)
+	d2 := lastRefresh.Add(maxPostponement).Sub(now)
+	if d1 < d2 {
+		return d1
+	}
+	return d2
+}
+
+// HoldRefresh marks affectingSnaps as held for refresh for up to holdTime.
+// HoldTime of zero denotes maximum allowed hold time.
+// Holding may fail for only some snaps in which case HoldError is returned and
+// it contains the details of failed ones.
+func HoldRefresh(st *state.State, gatingSnap string, holdDuration time.Duration, affectingSnaps ...string) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+	herr := &HoldError{
+		SnapsInError: make(map[string]HoldDurationError),
+	}
+	now := timeNow()
+	for _, heldSnap := range affectingSnaps {
+		hold, ok := gating[heldSnap][gatingSnap]
+		if !ok {
+			hold = &holdState{
+				FirstHeld: now,
+			}
+		}
+
+		lastRefreshTime, err := lastRefreshed(st, heldSnap)
+		if err != nil {
+			return err
+		}
+
+		maxDur := maxAllowedPostponement(gatingSnap, heldSnap)
```

Comment:
because maxPostponement-maxPostponementBuffer is anyway computed below,  maybe this function should take it explicit